### ISSUE TITLE
[BUG] GDN prefill adapter duplicates recurrent state pools, causing startup HBM OOM at higher request capacity

### ISSUE THREAD

**Issue by Yanko-7** [sgl-project/sglang-jax#1667]
## Problem

The fused GDN prefill adapter can introduce full recurrent-state-pool copies despite outer buffer donation. On Qwen3.5-27B this causes a startup XLA HBM OOM as the configured request/state-pool capacity grows, before any client requests arrive.

Observed while benchmarking #1658 rebased onto upstream main `0c74a5504e9986ae6fdd5890d7a546978c7f48d4`. The adapter is unchanged by that PR. 

## Reproduction

- Qwen/Qwen3.5-27B, TPU v7x-8 (4 chips / 8 JAX devices, topology 2x2x1), TP=8 / DP=4.
- JAX/jaxlib 0.11.1, libtpu 0.0.46.1, Flax 0.12.9, transformers 4.57.6; install the multimodal and TPU extras.

```bash
python -m sgl_jax.launch_server   --model-path "$MODEL_PATH" --trust-remote-code --skip-server-warmup   --device tpu --tp-size 8 --dp-size 4   --dtype bfloat16 --kv-cache-dtype bf16   --context-length 2048 --max-seq-len 2048   --max-running-requests 1024 --max-prefill-tokens 16384   --chunked-prefill-size 4096 --mem-fraction-static 0.9   --page-size 128 --disable-radix-cache --vision-encoder-parallel dp   --mm-processor-worker-num 2 --random-seed 0
```

The effective request limit becomes 768 because of the automatic recurrent pool's snapshot headroom. The first EXTEND precompile (`bs=768`, `tokens=1024`) fails with approximately **102.59 GiB required vs 94.74 GiB available**.

Expected: avoid unnecessary whole-pool duplication while preserving active/inactive/dummy state semantics, or correctly budget unavoidable scratch space before selecting cache capacities.

## Compiler evidence

The error's “HLO temporaries” wording is misleading here: 102.59 GiB is the **total default-memory allocation**, including parameters and aliased cache inputs, not an additional scratch allocation.

| Per-device allocation | GiB |
|---|---:|
| KV pools | 40.847656 |
| Weights and other inputs | 23.869160 |
| Recurrent state inputs (48 layers) | 18.070313 |
| Convolution state inputs | 0.362549 |
| Temporary arena | 19.393402 |
| Other | 0.044940 |
| Total | 102.588020 |

The temporary arena contains 48 full-pool copies of `f32[257,24,128,128]`, each **385.5 MiB**. Their distinct, nonoverlapping reserved address ranges total **18.0703125 GiB**, about 93.2% of the temporary arena. This is address-space attribution, not a claim that all copies are simultaneously live; other logical buffers can reuse those ranges at other schedule points.

Example optimized HLO: `copy.6367` copies `parameter766` (`memory_pools[0][0][1]`); `scatter_fusion.5` updates the copy for the fused kernel, and `scatter_fusion.7` later updates the original donated pool from the kernel result.

Relevant code:

- [Adapter initial-state gather/masking and temporary kernel-pool construction](https://github.com/sgl-project/sglang-jax/blob/0c74a5504e9986ae6fdd5890d7a546978c7f48d4/python/sgl_jax/srt/kernels/gdn/fused_chunk_parallel_adapter.py#L243).
- [Adapter result gather and scatter back to the original pool](https://github.com/sgl-project/sglang-jax/blob/0c74a5504e9986ae6fdd5890d7a546978c7f48d4/python/sgl_jax/srt/kernels/gdn/fused_chunk_parallel_adapter.py#L294).
- [Automatic recurrent pool sizing from max_running_requests](https://github.com/sgl-project/sglang-jax/blob/0c74a5504e9986ae6fdd5890d7a546978c7f48d4/python/sgl_jax/srt/model_executor/model_runner_kv_cache_mixin.py#L417).

A higher configured concurrency enlarges both the resident state pool and these compiler-generated copies. Fixed `mem_fraction_static` trades recurrent-state residency against KV capacity, but does not account for this additional copy arena. Outer donation alone does not remove copies when the graph still consumes the original pool.

## Focused validation of a possible fix

An **unsubmitted experimental no-tracking path** passes the donated state pools directly into the existing fused kernel and returns the updated pools. It preserves initial-state gating using effective sequence lengths; tracking retains the original path. No patch is included in #1658.

Single-layer real-TPU A/B, DP=4/TP=2, global tokens=1024, mixed new/resumed requests:

| Global pool / padded batch | Baseline median ms | Candidate median ms | Baseline temp MiB/device | Candidate temp MiB/device |
|---|---:|---:|---:|---:|
| 256 / 192 | 1.6748 | 0.9606 | 261.02 | 12.18 |
| 512 / 384 | 2.5496 | 1.1075 | 503.62 | 8.28 |
| 1024 / 768 | 4.5743 | 1.7141 | 999.47 | 31.16 |

All 9 cases (3 pool sizes × new/resumed/mixed requests) had zero maximum absolute difference in output, conv pool, and recurrent pool; inactive/dummy slots remained unchanged. The two full recurrent-pool copy instructions present in each single-layer baseline HLO disappeared in the candidate. Measurements exclude compilation/state initialization, with 40 synchronized samples per variant/case. Pool and padded batch vary together, so cross-size trends are not a pool-only ablation.

Remaining validation: tracking/snapshot semantics, speculative paths, multi-step state trajectories, and full-model logits/output equivalence. Do not extrapolate single-layer temporary-memory savings by multiplying by the layer count. Full-model peak memory after the candidate was not profiled.
